compat/winansi.c

```diff
@@ -573,6 +573,9 @@ static void detect_msys_tty(int fd)
 	if (!NT_SUCCESS(NtQueryObject(h, ObjectNameInformation,
 			buffer, sizeof(buffer) - 2, &result)))
 		return;
+	if (result <= sizeof(*nameinfo) || !nameinfo->Name.Buffer ||
```

Comment:
Please use `<`, not `<=`. The result could still be valid if the Buffer points outside the struct.

Reply:
I think `<=` would be correct, because in the case of `ObjectNameInformation` with no name, wine has
```c
        OBJECT_NAME_INFORMATION *p = ptr;
...
                if (!reply->total)  /* no name */
                {
                    if (sizeof(*p) > len) status = STATUS_INFO_LENGTH_MISMATCH;
                    else memset( p, 0, sizeof(*p) );
                    if (used_len) *used_len = sizeof(*p);
                }
```

and in that path, `used_len`, or `result` would be set to `sizeof(OBJECT_NAME_INFORMATION)`, or 16

If I use `result < sizeof(*nameinfo)`, the condition would never trigger as in all valid success paths, `used_len` is set to a minimum of `sizeof(*p)`. I do not think this would be related to `Buffer`'s location since `result` shouldn't have anything related to the pointer's location

Reply:
> If I use result < sizeof(*nameinfo), the condition would never trigger as in all valid success paths, used_len is set to a minimum of sizeof(*p). 

Are we sure that's necessarily true for non-Wine implementations, though? And wouldn't the `!nameinfo->Name.Buffer` part or the `!nameinfo->Name.Length` part still apply on wine?

Reply:
Yeah, that is true, just for completion’s sake, I've checked reactos' code for the function and looking at <https://github.com/reactos/reactos/blob/74513a75ab15608f2018f52661e84e8435b72b98/ntoskrnl/ob/obname.c#L1207>

they have something similar to
```c
    if (!(LocalInfo) || !(LocalInfo->Name.Buffer))
...
            *ReturnLength = sizeof(OBJECT_NAME_INFORMATION);
```

of course, I don't have access to the source code inside Windows, so I won't say anything about its behavior.


> And wouldn't the !nameinfo->Name.Buffer part or the !nameinfo->Name.Length part still apply on wine?

Those two conditions would still trigger, leading to the whole if being triggered, so it would be fine, I was just talking specifically about the first condition.

Rather, I could just remove the first condition and only check for `!nameinfo->name.Buffer || !nameinfo->Name.Length` to check the resulting information instead of caring about the reported information, would that be preferable?

Reply:
> I would, but if I did, I couldn't contribute code to GPL'ed software anymore.

That is true, and we really wouldn't want that

> ... For that reason, I ask again to replace the <= with a <.

That makes sense, done